### ISSUE TITLE
Create `How to develop PennyLane for 4 different ML frameworks` page

### ISSUE THREAD

**Issue by antalszava** [PennyLaneAI/pennylane#2943]
### Feature details

There are many subtleties that come up when developing the different Machine Learning frameworks that PennyLane supports. It would be great to add a new page in the developer documentation that notes these details.

1. Trainability: how does each framework mark trainability
A) Autograd: no way to mark, `argnum` argument can be passed to `autograd.grad`/`autograd.jacobian`; PennyLane patches Autograd's NumPy to allow marking tensors trainable via `requires_grad`.
B) Torch: tensors can be marked as trainable via `requires_grad`.
C) TensorFlow: only tracks trainability if being watched by a gradient tape. Although, there's a `trainable` argument that can be passed to `tf.Variable`, it is `True` by default and such tensors still require a gradient tape for differentiation.
D) JAX: no way to mark, `argnums` argument can be passed to `jax.grad`/`jax.jacobian`.

2. Writing tests:
* Naively thinking about each framework first implies creating 4 or more different test cases for each framework. Although one could leverage `pytest`'s parameterization, creating separate test cases may still be arguably the cleanest solution because of how different each framework's UI is.
* When computing gradients with TensorFlow, Jacobians should only be computed if absolutely necessary for the use case due to a potentially degraded performance with for example `diff_method="parameter-shift"` (see `Gotchas` section).

3. Gotchas:
* It was found that TensorFlow's `tape.jacobian` syntax yields poor performance with its default `experimental_use_pfor=True` when using `diff_method="parameter-shift"` (see https://github.com/PennyLaneAI/pennylane/pull/1869).
* For higher-order derivatives with TensorFlow, `experimental_use_pfor=False` has to be set for each "internal" `tapeN.jacobian` call with the "outermost" call allowing both allowing `experimental_use_pfor=False` or `experimental_use_pfor=True`.

E.g.,
```py
        with tf.GradientTape() as tape1:
            with tf.GradientTape(persistent=True) as tape2:
                res = circuit(x)
            g = tape2.jacobian(res, x, experimental_use_pfor=False) # <--- Needs experimental_use_pfor=False

        hess = tape1.jacobian(g, x) # <--- Can use experimental_use_pfor=True
```

### How important would you say this feature is?

2: Somewhat important. Needed this quarter.
